pkg/kubelet/cm/cpumanager/topology/topology_test.go

```diff
@@ -204,3 +205,383 @@ func Test_Discover(t *testing.T) {
 		})
 	}
 }
+
+func TestCPUDetailsKeepOnly(t *testing.T) {
+
+	var details CPUDetails
+	details = map[int]CPUInfo{
+		0: {},
+		1: {},
+		2: {},
+	}
+
+	tests := []struct {
+		name string
+		cpus cpuset.CPUSet
+		want CPUDetails
+	}{{
+		name: "cpus is in CPUDetails.",
+		cpus: cpuset.NewCPUSet(0, 1),
+		want: map[int]CPUInfo{
+			0: {},
+			1: {},
+		},
+	}, {
+		name: "cpus is not in CPUDetails.",
+		cpus: cpuset.NewCPUSet(3),
+		want: CPUDetails{},
+	}}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := details.KeepOnly(tt.cpus)
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("KeepOnly() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCPUDetailsNUMANodes(t *testing.T) {
+
+	tests := []struct {
+		name    string
+		details CPUDetails
+		want    cpuset.CPUSet
+	}{{
+		name: "Get CPUset of NUMANode IDs",
+		details: map[int]CPUInfo{
+			0: {NUMANodeID: 0},
+			1: {NUMANodeID: 0},
+			2: {NUMANodeID: 1},
+			3: {NUMANodeID: 1},
+		},
+		want: cpuset.NewCPUSet(0, 1),
+	}}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := tt.details.NUMANodes()
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("NUMANodes() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCPUDetailsNUMANodesInSockets(t *testing.T) {
+
+	var details CPUDetails
+	details = map[int]CPUInfo{
+		0: {SocketID: 0, NUMANodeID: 0},
+		1: {SocketID: 0, NUMANodeID: 1},
```

Comment:
I guess the same SocketID should be on the same NUMANodeID in a real machine, because physical CPU sockets are on each numa node.

Reply:
Thanks for your review.

NUMANodesInSockets returns all of the NUMANodeIDs associated with the given SocketIDs.
For example, if the SocketID are associated with two NUMANodeIDs, it returns two NUMANodeIDs.
I chose this value to test this behavior.
Should I change that the same SocketID only have the same NUMANodeID?

Reply:
> if the SocketID are associated with two NUMANodeIDs

Is the above situation common?
That means physical CPU socket on different nodes.

Reply:
> Is the above situation common?
> That means physical CPU socket on multiple NUMA nodes.

In the Linux kernel, NUMA nodes were associated with the CPU cores.
So, I think that physical CPU socket may be associated with multiple NUMA nodes.

Reference information:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/x86/mm/numa.c?h=v5.12.2
```
/*
 * There are unfortunately some poorly designed mainboards around that
 * only connect memory to a single CPU. This breaks the 1:1 cpu->node
 * mapping. To avoid this fill in the mapping for all possible CPUs,
 * as the number of CPUs is not known yet. We round robin the existing
 * nodes.
 */
static void __init numa_init_array(void)
{
	int rr, i;	rr = first_node(node_online_map);
	for (i = 0; i < nr_cpu_ids; i++) {
		if (early_cpu_to_node(i) != NUMA_NO_NODE)
			continue;
		numa_set_node(i, rr);
		rr = next_node_in(rr, node_online_map);
	}
}
```

Reply:
Thanks for your explanation, I see the `poorly designed mainboards` at the above comment lines ;-)

One comment based on your explanation: Current following test pattern is based on `poorly designed mainboards` case:
```
details = map[int]CPUInfo{
  0: {SocketID: 0, NUMANodeID: 0},
  1: {SocketID: 0, NUMANodeID: 1},
  2: {SocketID: 1, NUMANodeID: 2},
  3: {SocketID: 2, NUMANodeID: 3},
}
```
How about having 2 separated patterns like
```
details = map[int]CPUInfo{
  0: {SocketID: 0, NUMANodeID: 0},
  1: {SocketID: 1, NUMANodeID: 0},
  2: {SocketID: 2, NUMANodeID: 1},
  3: {SocketID: 3, NUMANodeID: 1},
}
```
and
```
details = map[int]CPUInfo{
  0: {SocketID: 0, NUMANodeID: 0},
  1: {SocketID: 0, NUMANodeID: 1},
  2: {SocketID: 1, NUMANodeID: 2},
  3: {SocketID: 1, NUMANodeID: 3},
}
```
?
Then last test pattern would be better to contain your explanation (`poorly designed mainboards` etc.) as a comment to avoid confusion in the future.

Reply:
I updated. Could you check it?

Reply:
In general, I saw architectures with multi NUMA per socket, NUMA mostly referred to the memory controller and not to the socket, please see - https://developer.amd.com/wp-content/resources/56338_1.00_pub.pdf, so IMHO it a little bi harsh comment :)

Reply:
In general it can go both ways. There can be multiple sockets-per-numa OR multiple-numa-per-socket. In fact, multiple-numa-.per-socket is starting to become quite common (please see something like the AMD Rome architecture for reference): https://hpcadvisorycouncil.atlassian.net/wiki/spaces/HPCWORKS/pages/1280442391/AMD+2nd+Gen+EPYC+CPU+Tuning+Guide+for+InfiniBand+HPC